### PR TITLE
feat: Persistent directories for Kubernetes

### PR DESCRIPTION
**Pull request by gbouv**
## Description:
Adds the docker backend to support persistent directories for user services. For now, we use `hostPath` persistent volumes, which will break if the user service is restarted onto a different k8s node than what it was before. There's plenty of ways to fix that (using k8s `local` persistent volumes, re-scheduling the pod on the same node as where it was before, using a shared block store like EBS on AWS, etc) but we are lacking concrete use cases to choose the best option right now. 
`hostPath` works for now and is the simplest implementation.

## Is this change user facing?
NO

### CODE REVIEW

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+		volumeName := volumeAttrs.GetName().GetString()
+		volumeLabelsStrs := map[string]string{}
+		for key, value := range volumeAttrs.GetLabels() {
+			volumeLabelsStrs[key.GetString()] = value.GetString()
+		}
+
+		var persistentVolume *apiv1.PersistentVolume
+		if persistentVolume, err = kubernetesManager.GetPersistentVolume(ctx, volumeName); err != nil {
+			persistentVolume, err = kubernetesManager.CreatePersistentVolume(ctx, namespace, volumeName, volumeLabelsStrs)
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go:99]: sorry, but I don't understand the logic here, I think that I'm missing something but, why the persistent volume is created if it already exists? does it work?

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+			if err != nil {
+				return nil, stacktrace.Propagate(err, "An error occurred creating the persistent volume for '%s'", persistentKey)
+			}
+			volumesCreated[persistentVolume.Name] = persistentVolume
+		}
+
+		// For now, we have a 1:1 mapping between volume and volume claims, so it's fine giving it the same name
+		var persistentVolumeClaim *apiv1.PersistentVolumeClaim
+		if persistentVolumeClaim, err = kubernetesManager.GetPersistentVolumeClaim(ctx, namespace, volumeName); err != nil {
+			persistentVolumeClaim, err = kubernetesManager.CreatePersistentVolumeClaim(ctx, namespace, volumeName, volumeLabelsStrs)
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go:109]: same comment here

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go [hunk 3]
```diff
@@ -78,6 +83,11 @@ var commandToRunWhenCreatingUserServiceShell = []string{
 }
 
 var (
+	volumeStorageClassName = "kurtosis-local-storage"
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go:91]: shouldn't it be a constant?

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go [hunk 4]
```diff
@@ -276,113 +293,280 @@ func (manager *KubernetesManager) GetServicesByLabels(ctx context.Context, names
...
+	//  the directory to already exist on the host. There's some k8s extensions to have them support dynamic provisioning
+	//  but we can't assume all Kubernetes cluster users will use will have those extensions. At least for now
+	//  - If this use case is hit only in the cloud (which is quite likely since having a multi-nodes k8s cluster running
+	//  outside a cloud provider infra is quite rare), then maybe we should just use whatever the cloud provider has,
+	//  like EBS for AWS for example. Those support dynamic provisioning and everything via their respective CSI drivers
+	nodes, err := manager.kubernetesClientSet.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
 	if err != nil {
-		return nil, stacktrace.Propagate(err, "Failed to parse volume size in megabytes %d", volumeSizeInMegabytes)
+		return nil, stacktrace.Propagate(err, "An unexpected error occurred retrieving the list of Kubernetes nodes.")
+	} else if len(nodes.Items) > 1 {
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go:305]: nice!

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go [hunk 5]
```diff
@@ -276,113 +293,280 @@ func (manager *KubernetesManager) GetServicesByLabels(ctx context.Context, names
...
+		return nil, stacktrace.NewError("Using persistent volumes on Kubernetes with multiple nodes is currently " +
+			"not supported. Reach out to Kurtosis if you need this feature.")
 	}
 
-	persistentVolumeClaim := &apiv1.PersistentVolumeClaim{
+	hostPathPath := "/kurtosis-persistent-service-data/" + namespace + "/" + volumeName
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go]: could use a constant for `kurtosis-persistent-service-data`

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go [hunk 6]
```diff
@@ -276,113 +293,280 @@ func (manager *KubernetesManager) GetServicesByLabels(ctx context.Context, names
...
+	volumeClaimName string,
+) (*apiv1.PersistentVolumeClaim, error) {
+	volumesClient := manager.kubernetesClientSet.CoreV1().PersistentVolumeClaims(namespace)
+	volumeClaim, err := volumesClient.Get(ctx, volumeClaimName, globalGetOptions)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "An error occurred getting the persistent volume claim '%s' in namespace '%s'", volumeClaimName, namespace)
+	}
+	deletionTimestamp := volumeClaim.GetObjectMeta().GetDeletionTimestamp()
+	if deletionTimestamp != nil {
+		return nil, stacktrace.Propagate(err, "Persistent volume claim with name '%s' in namespace '%s' has been marked for deletion",
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_manager/kubernetes_manager.go:541]: nice!

**Review by leoporoli — APPROVED** [review]

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go (hunk 1 above)
Comment (gbouv) [reply container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go:99]: No we do create it if the `GetPersistentVolume ` returns an error here (`err != nil`), i.e. if it does NOT exist.

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go (hunk 2 above)
Comment (gbouv) [reply container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go:109]: And same reply :D

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go (hunk 1 above)
Comment (leoporoli) [reply container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/user_services_functions/persistent_directories.go:99]: Oh! you are right. I misread the condition, it's my mistake

**Pull request merged**